ql/src/java/org/apache/hadoop/hive/ql/optimizer/lineage/OpProcFactory.java

```diff
@@ -676,6 +683,92 @@ public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
     }
   }
 
+  /**
+   * PTF processor
+   */
+  public static class PTFLineage implements SemanticNodeProcessor {
+
+    @Override
+    public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx, Object... nodeOutputs) throws SemanticException {
+      // LineageCTx
+      LineageCtx lCtx = (LineageCtx) procCtx;
+
+      // The operators
+      @SuppressWarnings("unchecked")
+      PTFOperator op = (PTFOperator)nd;
+      Operator<? extends OperatorDesc> inpOp = getParent(stack);
+      lCtx.getIndex().copyPredicates(inpOp, op);
+
+      Dependency dep = new Dependency();
+      DependencyType new_type = DependencyType.EXPRESSION;
+      dep.setType(new_type);
+      // TODO: Fix this to a non-null value. This comment comes from the default implementation (TransformLineage)
+      dep.setExpr(null);
+
+      List<String> columns = new ArrayList<>();
+      PartitionedTableFunctionDef funcDef = op.getConf().getFuncDef();
+
+      if (funcDef.getPartition() != null) {
+        addAllMakeUniqueIfNotNull(columns, funcDef.getPartition().getExpressions().getFirst().getExprNode().getCols());
+      }
+      if (funcDef.getOrder() != null) {
+        addAllMakeUniqueIfNotNull(columns, funcDef.getOrder().getExpressions().getFirst().getExprNode().getCols());
+      }
+
+      if (!(funcDef.getTFunction() instanceof Noop)) {
+
+          if (funcDef instanceof WindowTableFunctionDef
+                  && ((WindowTableFunctionDef) funcDef).getWindowFunctions().getFirst().getArgs() != null) {
+
+              for (PTFExpressionDef arg : ((WindowTableFunctionDef) funcDef).getWindowFunctions().getFirst().getArgs()) {
+
+                  if (!(arg.getExprNode() instanceof ExprNodeConstantDesc)) {
+                      columns.addAll(arg.getExprNode().getCols());
```

Comment:
Should column uniqueness be enforced here? I saw we use `addAllMakeUniqueIfNotNull` a few lines earlier.
Could `columns` be a set instead of list ?

Reply:
Good idea. Let me check that.

Reply:
I changed it to Set.